File: Python/108_Convert_Sorted_Array_to_Binary_SearchTree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def sortedArrayToBST(self, nums):
        """
        :type nums: List[int]
        :rtype: Optional[TreeNode]
        """

        def dfs(nums, idxs, idxe, res):
            print(idxs,idxe)
            if idxs == idxe: # base
                res.val = nums[idxs]
                return

            dif = idxe - idxs
            if dif > 1:
                idxm = dif // 2 + idxs
                if res != None:
                    res.val = nums[idxm]
                    if (idxm - 1) - idxs >= 1:
                        res.left = TreeNode()
                        dfs(nums, idxs, (idxm-1), res.left)
                    if idxe - (idxm + 1) >= 1:
                        res.right = TreeNode()
                        dfs(nums, (idxm+1), idxe, res.right)
            else:
                res.val = nums[idxe]
                res.left = TreeNode()
                dfs(nums, idxs, idxe - 1, res.left)
            
            return
        
        res = TreeNode()
        dfs(nums, 0, len(nums) - 1, res)

        return res


    
# algorithm: find a middle array which can be root in each layer
# if num of array is less than 2, a bigger value becomes root in each layer
# use index and recursive function, maintain        

File: C/108_ConvertSortedArraytoBinarySearchTree.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
struct TreeNode* createNode(int num) {
    struct TreeNode* newTr = (struct TreeNode*)malloc(sizeof(struct TreeNode));
    newTr->val = num;
    newTr->left = NULL;
    newTr->right = NULL;

    return newTr;
}

struct TreeNode* sortedArray(int* nums, int start, int end) {
    if (start > end) return NULL;

    int mid = (start + end) / 2;
    struct TreeNode* newTr = createNode(nums[mid]);

    newTr->left = sortedArray(nums, start, mid-1);
    newTr->right = sortedArray(nums, mid+1, end);

    return newTr;
}

struct TreeNode* sortedArrayToBST(int* nums, int numsSize) {
    return sortedArray(nums, 0, numsSize-1);
}

/*
The goal of this problem is to conver an array to height-balanced binary search tree. 
To solve this, I use a binary search approach to find the middle element of the array,
which serves as the root node at each level of the tree.
In this approach, the binary search is applied recursively until all elements in the array are processed. 
If the number of the elements in the current subarray is zero, the function returns null. 
Otherwise, a new node is created with the middele element as its value. 
Two recursive calls are then made to process the left and right halves of the array, 
and thier return values are assigned to the left and right child nodes of the current root, respectively. 
*/
